### ISSUE TITLE
Add more submodule information (git super status)

### ISSUE THREAD

**Issue by dscole** [magit/magit#2705]
I have a repository with many submodules and I'd like to be able to get an overview (git-status) of each of the submodules. Unfortunately `git submodule status` doesn't give me the overview I'd really like. So what I currently do is run this (sorry it's a bit dense, I use an alias for it):
```
git submodule foreach --quiet 'printf "%-20s" $name; echo -n $(git status|head -2); echo -n $(git status| grep modified| head -1); echo' | sort -r -k 4 | sort -k 7
```
Which prints an output like this:
```
subrepo0  On branch master Your branch is ahead of 'origin/master' by 1 commit.
subrepo1  On branch new_flag_feature Your branch is ahead of 'origin/master' by 1 commit.
subrepo2  On branch new_flag_feature Your branch is ahead of 'origin/master' by 1 commit.
subrepo3  On branch master Your branch is ahead of 'origin/master' by 1 commit.
subrepo4  On branch master Your branch is ahead of 'origin/master' by 1 commit.
subrepo5  On branch master Your branch is behind 'origin/master' by 13 commits, and can be fast-forwarded.
subrepo6  On branch master Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
subrepo7  On branch master Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
subrepo8  On branch master Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
subrepo9  On branch master Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
subrepo10 On branch master Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
subrepo11 On branch master Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
subrepo12 On branch master Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
subrepo13 On branch master Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
subrepo14 On branch master Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
subrepo15 On branch master Your branch is behind 'origin/master' by 3 commits, and can be fast-forwarded.
subrepo16 On branch master Your branch is behind 'origin/master' by 4 commits, and can be fast-forwarded.
subrepo17 On branch master Your branch is behind 'origin/master' by 6 commits, and can be fast-forwarded.
subrepo18 On branch master Your branch is behind 'origin/master' by 6 commits, and can be fast-forwarded.
subrepo19 On branch master Your branch is behind 'origin/master' by 9 commits, and can be fast-forwarded.
subrepo20 On branch master Your branch is up-to-date with 'origin/master'.
subrepo21 On branch master Your branch is up-to-date with 'origin/master'.
subrepo22 On branch master Your branch is up-to-date with 'origin/master'.
subrepo23 On branch master Your branch is up-to-date with 'origin/master'.
subrepo24 On branch master Your branch is up-to-date with 'origin/master'.
subrepo25 On branch master Your branch is up-to-date with 'origin/master'.
subrepo26 On branch master Your branch is up-to-date with 'origin/master'.
subrepo27 On branch master Your branch is up-to-date with 'origin/master'.
subrepo28 On branch master Your branch is up-to-date with 'origin/master'.
subrepo29 On branch master Your branch is up-to-date with 'origin/master'.
subrepo30 On branch master Your branch is up-to-date with 'origin/master'.
subrepo31 On branch master Your branch is up-to-date with 'origin/master'.
subrepo32 On branch master Your branch is up-to-date with 'origin/master'.
subrepo33 On branch master Your branch is up-to-date with 'origin/master'.
subrepo34 On branch master Your branch is up-to-date with 'origin/master'.
subrepo35 On branch master Your branch is up-to-date with 'origin/master'.
subrepo36 On branch master Your branch is up-to-date with 'origin/master'.
subrepo37 On branch master Your branch is up-to-date with 'origin/master'.
subrepo38 On branch master Your branch is up-to-date with 'origin/master'.
subrepo39 On branch master Your branch is up-to-date with 'origin/master'.
subrepo40 On branch master Your branch is up-to-date with 'origin/master'.
subrepo41 On branch master Your branch is up-to-date with 'origin/master'.
subrepo42 On branch master Your branch is up-to-date with 'origin/master'
subrepo43 On branch meta/config Your branch is up-to-date with 'origin/meta/config'.
```
I think that for someone that has many submodules in which they constantly change branches, it's very useful to have such an overview. Also, in magit this can become even more powerful. When I run this on the shell I want this to be as concise as possible, so I print each submodule on a single line. In magit you could have the full status for each submodule, simply have it collapsed to a single line by default (so it looks like my output above). Then theoretically you could allow each subrepo to be a link, such that if you press on it, it takes you to that repo's magit-status.

I call this output git super status.

**Comment by npostavs** [comment]
There is actually already some code for this, it's just not enabled by default. Try evaluating

```elisp
(dolist (inserter '(magit-insert-modules-unpulled-from-upstream
                    magit-insert-modules-unpulled-from-pushremote
                    magit-insert-modules-unpushed-to-upstream
                    magit-insert-modules-unpushed-to-pushremote
                    magit-insert-submodules))
  (magit-add-section-hook
   'magit-status-sections-hook inserter
   'magit-insert-unpulled-from-upstream))
```

You may also want to set `diff.submodule` in your git config, when set, magit prints more info about modified submodules. I'm also working on some submodule improvements, see #2674.

**Comment by dscole** [comment]
I tried it and it's pretty cool, I'll start using it immediately.

One issue I have with it though is that in the modules section I would like it to specify the branch each submodule is on. Having the tag is actually more useful than I would have thought, but I also need the branch. Another issue is that I would have like to have all the info (tag, branch, commits behind, commits ahead, etc...) in one table (like in my output), but I can live perfectly well with the approach you took.
Lastly, if you do end up adding the branch information to the submodules, it would be useful to be able to sort the modules by the branch/tag name. That way I see clearly all the modules which are currently on a dev branch (as opposed to a release branch), because I can put those at the top using sorting. Same goes for tags.

**Comment by npostavs** [comment]
> One issue I have with it though is that in the modules section I would like it to specify the branch each submodule is on.

Yes, that is the first improvement in #2674.

> Another issue is that I would have like to have all the info (tag, branch, commits behind, commits ahead, etc...) in one table (like in my output),

Sounds doable, but I would probably only start looking at how to do it after #2674 is finished.

> Lastly, if you do end up adding the branch information to the submodules, it would be useful to be able to sort the modules by the branch/tag name.

Hmm, right. I didn't think of this because I don't have any repositories with so many submodules. I wonder how it should be configured though (from a UI perspective).

**Comment by tarsius** [comment]
>> Lastly, if you do end up adding the branch information to the submodules, it would be useful to be able to sort the modules by the branch/tag name.

> Hmm, right. I didn't think of this because I don't have any repositories with so many submodules. I wonder how it should be configured though (from a UI perspective).

While not impossible, doing this in a regular magit buffer would be a huge hack. But we don't have to do it in a magit buffer, we can use a `tabulate-list-mode` buffer (like e.g. `package.el`) instead. Luckily I have already implemented that.

Currently that implementation only supports listing "all" repositories as specified using the option `magit-repository-directories` but it is easy to implement a submodule variant:

```lisp
(defvar magit-modulelist-columns magit-repolist-columns) ; TODO option

;;;###autoload
(defun magit-list-submodules ()
  "Display a list of the current repositories submodules."
  (interactive)
  (with-current-buffer (get-buffer-create "*Magit Submodules*") ; TODO per superrepo
    (magit-repolist-mode)
    (setq tabulated-list-entries
          (mapcar (lambda (path)
                    (let ((default-directory
                            (file-name-as-directory (expand-file-name path))))
                      (list path
                            (vconcat (--map (or (funcall (nth 2 it) path) "")
                                            magit-submodulelist-columns)))))
                  (magit-get-submodules)))
    (tabulated-list-print)
    (switch-to-buffer (current-buffer))))
```

Coincidently I implemented `magit-list-repositories` because a user requested something like the *Modules* section but for "all repositories, not just submodules". Now it is time to bring that feature back to be used with submodules ;-)

**Comment by tarsius** [comment]
>> One issue I have with it though is that in the modules section I would like it to specify the branch each submodule is on.

> Yes, that is the first improvement in #2674.

We can still do that anyway, if you wish, but I don't think we should go any further with regards to what information is being displayed.
